common/functions/src/scalars/comparisons/comparison.rs

```diff
@@ -231,22 +259,28 @@ pub trait ComparisonImpl: Sync + Send + Clone + 'static {
         M: PrimitiveType;
 
     fn eval_binary(_l: &[u8], _r: &[u8], _ctx: &mut EvalContext) -> bool;
+
+    fn eval_variant(_l: &VariantValue, _r: &VariantValue, _ctx: &mut EvalContext) -> bool;
 }
 
 pub trait ComparisonExpression: Sync + Send {
     fn eval(&self, l: &ColumnWithField, r: &ColumnWithField) -> Result<BooleanColumn>;
 }
 
 pub struct ComparisonScalarImpl<L: Scalar, R: Scalar, F> {
+    least_supertype: DataTypeImpl,
+    need_cast: bool,
     func: F,
     _phantom: PhantomData<(L, R)>,
 }
 
 impl<L: Scalar, R: Scalar, F> ComparisonScalarImpl<L, R, F>
 where F: Fn(L::RefType<'_>, R::RefType<'_>, &mut EvalContext) -> bool
 {
-    pub fn new(func: F) -> Self {
+    pub fn new(least_supertype: DataTypeImpl, need_cast: bool, func: F) -> Self {
```

Comment:
why we need `need_cast`, if the least_supertype is equal to the input type, we need cast.